File: pi-tidy-tools/packages/pi-tidy-bots/examples/community-python/backend.py
#!/usr/bin/env python3
"""Installed external backend fixture. No TypeScript or gateway source imports."""
import asyncio
import json
import os
import select
from pathlib import Path
import sys
from uuid import uuid4

sys.path.insert(0, str(Path(__file__).parent / "sdk"))
from tidy_backend_sdk import run_plugin

mode = "normal"
owned = None
launch_id = None
control_fd = None
attached_pid = None
pending_submits = {}


def record(ctx, kind, **values):
    path = Path(ctx.initialization["dataDir"]) / "native-calls.jsonl"
    with path.open("a", encoding="utf-8") as stream:
        stream.write(json.dumps({"kind": kind, **values}) + "\n")
        stream.flush()
        os.fsync(stream.fileno())


def write_raw_frame(payload):
    """Write adversarial transport bytes without SDK framing/validation."""
    view = memoryview(payload)
    fd = sys.stdout.fileno()
    while view:
        try:
            written = os.write(fd, view)
            view = view[written:]
        except BlockingIOError:
            select.select([], [fd], [], 1.0)
    sys.stdout.flush()


async def initialize(config, ctx):
    global mode, attached_pid
    mode = config.get("mode", "normal")
    if mode == "attached":
        ctx.ownership = "attached"
        attached_pid = config["externalPid"]
    record(ctx, "initialized")


async def opened(p, ctx):
    global owned, launch_id, control_fd
    record(ctx, "open", openId=p["openId"])
    if mode == "crash-open":
        os._exit(17)
    if mode == "owned":
        owned = await asyncio.create_subprocess_exec(sys.executable, "-c", "import time; time.sleep(600)")
        record(ctx, "owned", pid=owned.pid)
    if mode == "registered-owned":
        launch_id = "tidy-launch-" + str(uuid4())
        prepared = await ctx.owned_process("prepare", {"launchId": launch_id})
        if prepared.get("launcherProtocol") != 2:
            raise ValueError("Unsupported launcher protocol")
        read_fd, control_fd = os.pipe()
        effect = str(Path(ctx.initialization["dataDir"]) / "child-effect")
        try:
            owned = await asyncio.create_subprocess_exec(
                prepared["executable"], prepared["launcherPath"], launch_id, "--control-fd=" + str(read_fd),
                sys.executable, "-I", "-c", "from pathlib import Path;import time;Path(" + repr(effect) + ").write_text('started');time.sleep(600)",
                env={}, pass_fds=(read_fd,), start_new_session=True,
                stdin=asyncio.subprocess.DEVNULL, stdout=asyncio.subprocess.DEVNULL, stderr=asyncio.subprocess.DEVNULL)
            admitted = await ctx.owned_process("record", {"launchId": launch_id, "pid": owned.pid})
            if admitted.get("state") != "started" or admitted.get("identity", {}).get("pid") != owned.pid:
                raise ValueError("Missing ownership receipt")
            os.write(control_fd, (json.dumps({"activate": launch_id, "env": {}}) + "\n").encode())
            record(ctx, "registered", launchId=launch_id, pid=owned.pid)
        finally:
            os.close(read_fd)
    return {"status": "opened", "nativeReference": "python:" + p["openId"]}


async def submit(p, ctx):
    global pending_submits
    record(ctx, "submit", operationId=p["operationId"], text=p["input"][0]["text"])
    if mode == "crash-submit":
        os._exit(18)
    if mode == "marker-crash":
        record(ctx, "malformed", mode=mode, operationId=p["operationId"])
        os._exit(19)
    if mode in ("cancel-ack", "cancel-delayed", "cancel-lost"):
        pending_submits[p["operationId"]] = {"params": p, "cancelled": asyncio.Event()}
        try:
            await ctx.emit({"operationId": p["operationId"], "turnId": p["turnId"], "type": "turn.started", "payload": {}})
            await pending_submits[p["operationId"]]["cancelled"].wait()
        finally:
            pending_submits.pop(p["operationId"], None)
        return {"disposition": "accepted"}
    if mode in ("malformed-json", "malformed-event", "oversize", "nonfinite"):
        record(ctx, "malformed", mode=mode, operationId=p["operationId"])
        if mode == "malformed-json":
            raw = b'{"jsonrpc":"2.0","method":"event","params":\n'
        elif mode == "malformed-event":
            raw = (json.dumps({"jsonrpc": "2.0", "method": "event", "params": {
                "type": "not-a-protocol-event", "payload": {}}}) + "\n").encode()
        elif mode == "nonfinite":
            raw = b'{"jsonrpc":"2.0","method":"event","params":{"type":"turn.started","payload":{"value":NaN}}}\n'
        else:
            raw = (b'{"jsonrpc":"2.0","method":"event","params":{"type":"turn.started","payload":{"padding":"' +
                   (b"x" * (2 * 1024 * 1024)) + b'"}}}\n')
        write_raw_frame(raw)
        # Exercise the host's post-fault isolation path; this must not become
        # a successful canonical completion if the host has closed the pipe.
        await ctx.emit({"operationId": p["operationId"], "turnId": p["turnId"], "type": "turn.started", "payload": {}})
    if mode == "timeout":
        try:
            await asyncio.sleep(60)
        except asyncio.CancelledError:
            record(ctx, "handler_cancelled", operationId=p["operationId"])
            raise
    if mode == "reverse":
        await ctx.host_call("operator.enqueue", {"title": p["input"][0]["text"]}, operation_id=p["operationId"], tool_call_id="tool-1", action_id="action-1", payload_digest="sha256:reverse")
    ids = {"operationId": p["operationId"], "turnId": p["turnId"]}
    await ctx.emit({**ids, "type": "turn.started", "payload": {}})
    message = {**ids, "messageId": "message:" + p["operationId"]}
    await ctx.emit({**message, "type": "message.started", "payload": {"role": "assistant", "order": 0}})
    text = "Python: " + p["input"][0]["text"]
    await ctx.emit({**message, "blockId": "body", "type": "text.snapshot", "payload": {"revision": 1, "text": text}})
    await ctx.emit({**message, "type": "message.finished", "payload": {"ts": "2026-09-05T12:00:00.000Z", "blocks": [{"type": "text", "blockId": "body", "revision": 1, "text": text}]}})
    await ctx.emit({**ids, "type": "turn.terminal", "payload": {"execution": "ended", "observation": "complete"}})
    return {"disposition": "accepted"}


async def cancelled(p, ctx):
    global pending_submits
    record(ctx, "cancel", operationId=p["operationId"], targetOperationId=p["targetOperationId"])
    target = pending_submits.get(p["targetOperationId"])
    if mode == "cancel-lost":
        os._exit(20)
    if mode == "cancel-ack" and target is not None:
        ids = {"operationId": target["params"]["operationId"], "turnId": target["params"]["turnId"]}
        await ctx.emit({**ids, "type": "turn.terminal", "payload": {"execution": "cancelled", "observation": "complete"}})
        target["cancelled"].set()
    return {"status": "requested"}


async def respond(p, ctx):
    record(ctx, "decision", operationId=p["operationId"], optionId=p["optionId"])
    return {"status": "applied"}


async def snapshot(p, ctx):
    return {"sourceSequence": ctx.store.sequence, "nativeOutcome": "unknown"}


async def closed(p, ctx):
    global control_fd
    if launch_id is not None:
        if control_fd is not None:
            os.close(control_fd)
            control_fd = None
        if owned is not None:
            await owned.wait()
        for method in ("prepare", "record"):
            try:
                await ctx.owned_process(method, {"launchId": launch_id})
                raise ValueError("Cleanup admitted native launch")
            except Exception as error:
                if getattr(error, "code", None) != "plugin_closed":
                    raise
                record(ctx, "cleanup_denied", method=method)
        inspected = await ctx.owned_process("inspect", {"launchId": launch_id})
        stopped = await ctx.owned_process("stopped", {"launchId": launch_id})
        if stopped.get("state") != "stopped":
            raise ValueError("Cleanup lacks ownership proof")
        record(ctx, "cleanup_confirmed", inspected=inspected["launchId"], stopped=stopped["launchId"])
        return {"ownedStopped": True}
    if attached_pid is not None:
        # Fixture deliberately makes the lifecycle contract observable against
        # a process started by the test outside the plugin ownership tree.
        if p["stopOwned"]:
            os.kill(attached_pid, 15)
        record(ctx, "attached_closed", stopOwned=p["stopOwned"], pid=attached_pid)
        return {"ownedStopped": True}
    if owned is not None and p["stopOwned"]:
        if owned.returncode is None:
            owned.terminate()
        await owned.wait()
        record(ctx, "reaped", pid=owned.pid)
        return {"ownedStopped": True}
    if ctx.initialization:
        record(ctx, "closed", stopOwned=p["stopOwned"], mode=p["mode"], ownership=p["ownership"])
    return {"ownedStopped": False}


run_plugin(
    identity={"id": "org.example.tidy-community-python", "version": "1.0.0"},
    runtime={"name": "community-python-example", "version": "1.0.0", "transport": "stdio"},
    capabilities={
        "input": {"text": True, "mediaTypes": [], "maxMediaBytes": 0},
        "sessions": {"load": False, "import": False, "continuity": "unverified"},
        "output": {"text": "snapshots", "tools": False, "usage": "unknown"},
        "operations": {"nativeDedupe": "none", "nativeReplay": "none", "cancel": "cooperative", "steer": False},
        "interactions": {"permissions": "exact-request", "questions": False},
        "configuration": {"model": False, "thinking": False, "compact": False}, "fleetTools": False,
    },
    handlers={"session.open": opened, "operation.submit": submit, "operation.cancel": cancelled,
              "interaction.respond": respond, "session.snapshot": snapshot},
    on_initialize=initialize, on_close=closed,
)

File: pi-tidy-tools/packages/pi-tidy-bots/test/fixtures/plugin-sdk/backend.mjs
#!/usr/bin/env node
import { appendFileSync, openSync, fsyncSync, closeSync } from "node:fs";
import { join } from "node:path";
import { spawn } from "node:child_process";
import { randomUUID } from "node:crypto";
import { isDeepStrictEqual } from "node:util";
import { runPlugin, spawnOwnedProcess } from "SDK_INDEX_URL";
const { HermesInteractions } = await import("HERMES_INTERACTIONS_URL");

let ownedChild;
let ownedLaunchId;
let ownedHandle;
let interactions;
const log = (value) => {
  const file = openSync(
    join(process.env.TIDY_DATA_DIR, "native-effects.jsonl"),
    "a"
  );
  try {
    appendFileSync(file, JSON.stringify(value) + "\n");
    fsyncSync(file);
  } finally {
    closeSync(file);
  }
};
const capabilities = {
  input: { text: true, mediaTypes: [], maxMediaBytes: 0 },
  sessions: { load: false, import: false, continuity: "unverified" },
  output: { text: "snapshots", tools: false, usage: "unknown" },
  operations: {
    nativeDedupe: "none",
    nativeReplay: "none",
    cancel: "cooperative",
    steer: false,
  },
  interactions: { permissions: "exact-request", questions: false },
  configuration: { model: true, thinking: false, compact: false },
  fleetTools: true,
};
const runtime = runPlugin({
  identity: { id: "org.example.sdk-fixture", version: "1.0.0" },
  runtime: { name: "sdk-native-fixture", version: "1.0.0" },
  capabilities,
  ownership: process.env.FIXTURE_OWNERSHIP ?? "owned",
  onInitialize(ctx) {
    if (ctx.initialization.config.mode?.startsWith("hermes-")) {
      interactions = new HermesInteractions(ctx, {
        timeoutMs:
          ctx.initialization.config.mode === "hermes-lost-receipt" ? 150 : 3000,
        onFailure: () => {
          void runtime.close("permission_observation_gap");
        },
      });
    }
    if (ctx.initialization.config.mode === "owned-child") {
      ownedChild = spawn(
        process.execPath,
        ["-e", "setInterval(() => {}, 1000)"],
        { stdio: "ignore" }
      );
      log({ childPid: ownedChild.pid });
    }
  },
  async onClose(info, ctx) {
    interactions?.close();
    log({
      close: info.reason,
      ownership: info.ownership,
      mode: info.mode,
      signalAborted: ctx.signal.aborted,
    });
    if (ctx.initialization.config.mode === "stuck-cleanup")
      await new Promise(() => {});
    if (ctx.initialization.config.mode === "cleanup-probe") {
      for (const method of ["prepare", "record", "inspect", "stopped"]) {
        try {
          await ctx.ownedProcess(method, {
            launchId: `tidy-launch-${randomUUID()}`,
          });
          throw new Error("Unexpected cleanup service admission");
        } catch (error) {
          if (error.code !== "parent_eof") throw error;
          log({ cleanupDenied: method });
        }
      }
    }
    if (
      ownedChild &&
      ownedChild.exitCode === null &&
      ownedChild.signalCode === null
    ) {
      const exit = new Promise((resolve) => ownedChild.once("close", resolve));
      if (ownedHandle) await ownedHandle.close();
      else ownedChild.kill("SIGTERM");
      await exit;
    }
    if (ownedLaunchId) {
      for (const method of ["prepare", "record"]) {
        try {
          await ctx.ownedProcess(method, {
            launchId: ownedLaunchId,
            ...(method === "record" ? { pid: ownedChild.pid } : {}),
          });
          throw new Error("Cleanup admitted a native launch");
        } catch (error) {
          if (error.code !== "parent_eof") throw error;
          log({ cleanupDenied: method });
        }
      }
      const inspected = await ctx.ownedProcess("inspect", {
        launchId: ownedLaunchId,
      });
      const stopped = await ctx.ownedProcess("stopped", {
        launchId: ownedLaunchId,
      });
      if (stopped.state !== "stopped")
        throw new Error("Cleanup lacks ownership proof");
      log({
        cleanupInspected: inspected.launchId,
        cleanupStopped: stopped.launchId,
      });
    }
    return { ownedResourcesStopped: true };
  },
  handlers: {
    async "session.open"(params, ctx) {
      log({ method: "session.open", openId: params.openId });
      if (ctx.initialization.config.mode === "registered-child") {
        ownedLaunchId = `tidy-launch-${randomUUID()}`;
        log({ ownedLaunchId });
        // A caller environment must never preload code in the trusted wrapper.
        process.env.NODE_OPTIONS = `--require=${join(ctx.initialization.dataDir, "preload.cjs")}`;
        ownedHandle = await spawnOwnedProcess(ctx, {
          launchId: ownedLaunchId,
          executable: process.execPath,
          args: [
            "-e",
            `require('node:fs').writeFileSync(${JSON.stringify(join(ctx.initialization.dataDir, "child-effect"))},'started');setInterval(()=>{},1000)`,
          ],
          cwd: ctx.initialization.workspace,
          environment: {},
        });
        ownedChild = ownedHandle.child;
        ownedChild.stdout.resume();
        ownedChild.stderr.resume();
      }
      if (ctx.initialization.config.mode === "crash-open")
        process.kill(process.pid, "SIGKILL");
      return { status: "opened", nativeReference: `native:${params.openId}` };
    },
    async "session.snapshot"(_params, ctx) {
      if (ownedLaunchId)
        return ctx.ownedProcess("inspect", { launchId: ownedLaunchId });
      return {
        disposition: "unknown",
        observation: ctx.store.observationGap
          ? "reconciliation_required"
          : "complete",
        lastSourceSequence: ctx.store.watermark,
      };
    },
    async "operation.submit"(params, ctx) {
      log({ method: "operation.submit", operationId: params.operationId });
      const text = params.input[0].text;
      if (text === "[crash]") process.kill(process.pid, "SIGKILL");
      if (text === "[hang]") await new Promise(() => {});
      if (text === "[slow]") {
        await new Promise((resolve) => setTimeout(resolve, 100));
        log({ drainedWithoutAbort: !ctx.signal.aborted });
      }
      if (
        text === "[host-action]" ||
        text === "[host-action-unknown]" ||
        text === "[host-action-reconcile]" ||
        text === "[host-action-reconcile-bad]"
      ) {
        const call = {
          name: "fleet.send",
          callId: "call-1",
          actionId: "action-1",
          operationId: params.operationId,
          toolCallId: "tool-1",
          payloadDigest: "caller-action-digest",
          arguments: { target: "fixture", text: "hello" },
        };
        const first = text.startsWith("[host-action-reconcile")
            ? (ctx.store.reserve(
                "action:action-1",
                "host.call",
                call.payloadDigest,
                call
              ),
              await ctx.reconcileHostAction("action-1"))
            : await ctx.hostCall(call),
          second = text.startsWith("[host-action-reconcile")
            ? first
            : await ctx.hostCall(call);
        log({
          actionResultsEqual: isDeepStrictEqual(first, second),
          actionResult: first,
        });
      }
      const identity = {
        operationId: params.operationId,
        turnId: params.turnId,
      };
      if (interactions)
        ctx.emit({
          ...identity,
          type: "operation.disposition",
          payload: { disposition: "accepted" },
        });
      ctx.emit({ ...identity, type: "turn.started", payload: {} });
      if (interactions) {
        const permissionId = "native-permission-1";
        const response = await interactions.onPermission(
          {
            _meta: { tidy: { permissionId } },
            options: [
              {
                optionId: "allow_once",
                kind: "allow_once",
                name: "Allow once",
              },
              { optionId: "deny", kind: "reject_once", name: "Deny" },
            ],
          },
          17,
          identity,
          ctx.signal
        );
        log({ nativePermissionResponse: response });
        if (
          response.outcome.outcome === "selected" &&
          ctx.initialization.config.mode === "hermes-confirmed"
        )
          interactions.onPermissionConsumed({
            ...identity,
            permissionId,
            optionId: response.outcome.optionId,
          });
      }
      ctx.emit({
        ...identity,
        messageId: `message:${params.operationId}`,
        type: "message.started",
        payload: { role: "assistant", order: 0 },
      });
      ctx.emit({
        ...identity,
        messageId: `message:${params.operationId}`,
        blockId: "body",
        type: "text.snapshot",
        payload: { revision: 1, text: "Reply: " + text },
      });
      ctx.emit({
        ...identity,
        messageId: `message:${params.operationId}`,
        type: "message.finished",
        payload: {
          ts: "2026-09-05T00:00:00.000Z",
          blocks: [
            {
              type: "text",
              blockId: "body",
              revision: 1,
              text: "Reply: " + text,
            },
          ],
        },
      });
      ctx.emit({
        ...identity,
        type: "turn.terminal",
        payload: { execution: "ended", observation: "complete" },
      });
      return { disposition: "accepted" };
    },
    "operation.cancel"(params) {
      log({ method: "operation.cancel", operationId: params.operationId });
      return { status: "requested" };
    },
    "interaction.respond"(params) {
      log({ method: "interaction.respond", operationId: params.operationId });
      if (interactions) return interactions.respond(params);
      return { status: "applied" };
    },
    "session.configure"(params) {
      log({ method: "session.configure", operationId: params.operationId });
      return { status: "applied" };
    },
  },
});
const outcome = await runtime.done;
log({ done: outcome });
process.exit(outcome.cleanup === "complete" ? 0 : 2);
